### PR TITLE
Introduce Product request/response interceptors

### PR DESCRIPTION
**Pull request by sav007**
Missing JUnit tests will add as reminder task for that.

@krisorr @yervantk

### CODE REVIEW

**Comment by yervantk** [comment]
LGTM

In future we should look into simplifying it, cause it's hard to understand, but right now I don't know what that would mean.

**Comment by krisorr** [comment]
@yervantk which parts did you find overcomplicated or hard to understand?

File: MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/ApiRequestInterceptWrapper.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+ *   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *   OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ *   THE SOFTWARE.
+ */
+package com.shopify.buy.dataprovider;
+
+import rx.Observable;
+import rx.functions.Func2;
+
+final class ApiRequestInterceptWrapper {
```
Comment (krisorr) [inline MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/ApiRequestInterceptWrapper.java]: should this just be ApiInterceptWrapper?

File: MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/ProductServiceDefault.java [hunk 2]
```diff
@@ -121,14 +144,29 @@ public CancellableTask getProduct(final Long productId, final Callback<Product>
...
             .doOnNext(new RetrofitSuccessHttpStatusCodeHandler<>())
             .compose(new UnwrapRetrofitBodyTransformer<ProductListings, List<Product>>())
             .compose(new FirstListItemOrDefaultTransformer<Product>())
-            .onErrorResumeNext(new BuyClientExceptionHandler<Product>())
-            .observeOn(callbackScheduler);
+            .onErrorResumeNext(new BuyClientExceptionHandler<Product>());
+
+        return ApiRequestInterceptWrapper.wrap(
+            apiRequest,
+            new ApiRequestInterceptWrapper.InterceptorCall<ProductApiInterceptor, Product>(requestInterceptor) {
```
Comment (krisorr) [inline MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/ProductServiceDefault.java]: something is off here.  Both the request and response do exactly the same thing - can we factor this out?

**Comment by krisorr** [comment]
LGTM 🚢

**Pull request merged**